nexus/reconfigurator/simulation/src/config.rs

```diff
@@ -0,0 +1,135 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use std::ops::Deref;
+
+use indexmap::IndexSet;
+use omicron_common::api::external::Name;
+
+use crate::errors::{DuplicateError, MissingError};
+
+/// Versioned simulator configuration.
+///
+/// This is part of the state that is versioned and stored in the store.
+#[derive(Clone, Debug)]
+pub struct SimConfig {
+    /// Set of silo names configured
+    ///
+    /// These are used to determine the contents of external DNS.
+    silo_names: IndexSet<Name>,
+
+    /// External DNS zone name configured
+    external_dns_zone_name: String,
+
+    /// The number of Nexus zones to create.
+    ///
+    /// TODO: This doesn't quite fit in here because it's more of a policy
+    /// setting than a config option. But we can't set it in the
+    /// `SystemDescription` because need to persist policy across system wipes.
+    /// So users have to remember to set num_nexus twice: once in the config
+    /// and once in the policy.
+    ///
+    /// We can likely make this better after addressing
+    /// https://github.com/oxidecomputer/omicron/issues/6803.
+    num_nexus: Option<u16>,
+}
+
+impl SimConfig {
+    pub(crate) fn new() -> Self {
+        Self {
+            // We use "example-silo" here rather than "default-silo" to make it
+            // clear that we're in a test environment.
+            silo_names: std::iter::once("example-silo".parse().unwrap())
+                .collect(),
+            external_dns_zone_name: String::from("oxide.example"),
+            num_nexus: None,
+        }
+    }
+
+    #[inline]
+    pub fn silo_names(&self) -> impl ExactSizeIterator<Item = &Name> {
+        self.silo_names.iter()
+    }
+
+    #[inline]
+    pub fn external_dns_zone_name(&self) -> &str {
+        &self.external_dns_zone_name
+    }
+
+    #[inline]
+    pub fn num_nexus(&self) -> Option<u16> {
+        self.num_nexus
+    }
+
+    pub(crate) fn to_mut(&self) -> MutableSimConfig {
+        MutableSimConfig { config: self.clone(), log: Vec::new() }
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct MutableSimConfig {
```

Comment:
TODO: Probably change this to `SimConfigBuilder` (and `SimSystemBuilder` etc) since that matches a lot of what we're doing.